## Importación de librerías

In [323]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

from toolbox_DS import *

## Funciones

## Cargo el dataset

In [324]:
df = pd.read_excel('./data/marketing_campaign.xlsx')
df.head(8)

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   
5  7446        1967      Master       Together  62513.0        0         1   
6   965        1971  Graduation       Divorced  55635.0        0         1   
7  6177        1985         PhD        Married  33454.0        1         0   

  Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
0  2012-09-04       58       635  ...                  7             0   
1  2014-03-08       38        11  ...                  5             0   
2  2013-08-21       26       426  ...                  4             0   
3  2014-02-10       26        11  ...                  6             0   
4  2014-01-19       94       173  ...                  5             0   
5  2013-09-09       16       520  ...                  6             0   
6  2012-11-13       34       235  ...                  6             0   
7  2013-05-08       32        76  ...                  8             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   
5             0             0             0             0         0   
6             0             0             0             0         0   
7             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0  
5              3         11         0  
6              3         11         0  
7              3         11         0  

[8 rows x 29 columns]

In [325]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

## Tabla con significado de variables

| Nombre de Columna       | Descripción                                           |
|-------------------------|-------------------------------------------------------|
| ID                      | Identificador único para cada cliente                 |
| Year_Birth              | Año de nacimiento del cliente                         |
| Education               | Nivel educativo del cliente                           |
| Marital_Status          | Estado civil del cliente                              |
| Income                  | Ingresos anuales del cliente                          |
| Kidhome                 | Número de niños pequeños en el hogar del cliente      |
| Teenhome                | Número de adolescentes en el hogar del cliente        |
| Dt_Customer             | Fecha en que el cliente se unió                       |
| Recency                 | Número de días desde la última compra                 |
| MntWines                | Cantidad gastada en vino en los últimos 2 años        |
| MntFruits               | Cantidad gastada en frutas en los últimos 2 años      |
| MntMeatProducts         | Cantidad gastada en carne en los últimos 2 años       |
| MntFishProducts         | Cantidad gastada en pescado en los últimos 2 años     |
| MntSweetProducts        | Cantidad gastada en dulces en los últimos 2 años      |
| MntGoldProds            | Cantidad gastada en oro en los últimos 2 años         |
| NumDealsPurchases       | Número de compras realizadas con descuento            |
| NumWebPurchases         | Número de compras realizadas a través del sitio web de la empresa |
| NumCatalogPurchases     | Número de compras realizadas utilizando un catálogo   |
| NumStorePurchases       | Número de compras realizadas directamente en tiendas  |
| NumWebVisitsMonth       | Número de visitas al sitio web de la empresa en el último mes |
| AcceptedCmp3            | 1 si el cliente aceptó la oferta en la 3ra campaña, 0 en caso contrario |
| AcceptedCmp4            | 1 si el cliente aceptó la oferta en la 4ta campaña, 0 en caso contrario |
| AcceptedCmp5            | 1 si el cliente aceptó la oferta en la 5ta campaña, 0 en caso contrario |
| AcceptedCmp1            | 1 si el cliente aceptó la oferta en la 1ra campaña, 0 en caso contrario |
| AcceptedCmp2            | 1 si el cliente aceptó la oferta en la 2da campaña, 0 en caso contrario |
| Complain                | 1 si el cliente se quejó en los últimos 2 años, 0 en caso contrario    |
| Z_CostContact           | Costo de contacto del cliente (valor constante)       |
| Z_Revenue               | Ingresos por el cliente (valor constante)             |
| Response                | 1 si el cliente aceptó la oferta en la última campaña, 0 en caso contrario |



## Vistazo general variables

In [326]:
describe_df(df).T

DATE_TYPE MISSINGS (%) UNIQUE_VALUES CARDIN (%)
ID                      int64          0.0          2240      100.0
Year_Birth              int64          0.0            59       2.63
Education              object          0.0             5       0.22
Marital_Status         object          0.0             8       0.36
Income                float64         1.07          1974      89.08
Kidhome                 int64          0.0             3       0.13
Teenhome                int64          0.0             3       0.13
Dt_Customer            object          0.0           663       29.6
Recency                 int64          0.0           100       4.46
MntWines                int64          0.0           776      34.64
MntFruits               int64          0.0           158       7.05
MntMeatProducts         int64          0.0           558      24.91
MntFishProducts         int64          0.0           182       8.12
MntSweetProducts        int64          0.0           177        7.9
MntGoldProds            int64          0.0           213       9.51
NumDealsPurchases       int64          0.0            15       0.67
NumWebPurchases         int64          0.0            15       0.67
NumCatalogPurchases     int64          0.0            14       0.62
NumStorePurchases       int64          0.0            14       0.62
NumWebVisitsMonth       int64          0.0            16       0.71
AcceptedCmp3            int64          0.0             2       0.09
AcceptedCmp4            int64          0.0             2       0.09
AcceptedCmp5            int64          0.0             2       0.09
AcceptedCmp1            int64          0.0             2       0.09
AcceptedCmp2            int64          0.0             2       0.09
Complain                int64          0.0             2       0.09
Z_CostContact           int64          0.0             1       0.04
Z_Revenue               int64          0.0             1       0.04
Response                int64          0.0             2       0.09

Analizando las variables por debajo del 17 en 'Unique values'

In [327]:
temp = describe_df(df).T 
val_card = temp.loc[temp['UNIQUE_VALUES'] < 17].index
print(val_card)
for cada in val_card:
    #print(cada)
    print(df[cada].value_counts(True))
    print()

Index(['Education', 'Marital_Status', 'Kidhome', 'Teenhome',
       'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases',
       'NumStorePurchases', 'NumWebVisitsMonth', 'AcceptedCmp3',
       'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2',
       'Complain', 'Z_CostContact', 'Z_Revenue', 'Response'],
      dtype='object')
Education
Graduation    0.503125
PhD           0.216964
Master        0.165179
2n Cycle      0.090625
Basic         0.024107
Name: proportion, dtype: float64

Marital_Status
Married     0.385714
Together    0.258929
Single      0.214286
Divorced    0.103571
Widow       0.034375
Alone       0.001339
Absurd      0.000893
YOLO        0.000893
Name: proportion, dtype: float64

Kidhome
0    0.577232
1    0.401339
2    0.021429
Name: proportion, dtype: float64

Teenhome
0    0.516964
1    0.459821
2    0.023214
Name: proportion, dtype: float64

NumDealsPurchases
1     0.433036
2     0.221875
3     0.132589
4     0.084375
5     0.041964
6     0.027

Después de este análisis se ha visto que hay valores en 'Marital_Status' un poco raros:
- 'YOLO' y 'Absurd'

In [328]:
df.loc[(df['Marital_Status'] == 'Absurd') | (df['Marital_Status']== 'YOLO' )]

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
2093   7734        1993  Graduation         Absurd  79244.0        0   
2134   4369        1957      Master         Absurd  65487.0        0   
2177    492        1973         PhD           YOLO  48432.0        0   
2202  11133        1973         PhD           YOLO  48432.0        0   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
2093         0  2012-12-19       58       471  ...                  1   
2134         0  2014-01-10       48       240  ...                  2   
2177         1  2012-10-18        3       322  ...                  8   
2202         1  2012-10-18        3       322  ...                  8   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
2093             0             0             1             1             0   
2134             0             0             0             0             0   
2177             0             0             0             0             0   
2202             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
2093         0              3         11         1  
2134         0              3         11         0  
2177         0              3         11         0  
2202         0              3         11         1  

[4 rows x 29 columns]

'YOLO' parece que se ha metido a propósito. Habrá que eliminarla.  
'Absurd' parece que hay que cambiarle el nombre ya que puede aportar informaicón útil.

## Valores Nulos

In [329]:
df.isna().sum()

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64

Hay 24 valores nulos en la variable 'Income'

In [330]:
df.loc[df['Income'].isna()]

ID  Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
10     1994        1983  Graduation        Married     NaN        1         0   
27     5255        1986  Graduation         Single     NaN        1         0   
43     7281        1959         PhD         Single     NaN        0         0   
48     7244        1951  Graduation         Single     NaN        2         1   
58     8557        1982  Graduation         Single     NaN        1         0   
71    10629        1973    2n Cycle        Married     NaN        1         0   
90     8996        1957         PhD        Married     NaN        2         1   
91     9235        1957  Graduation         Single     NaN        1         1   
92     5798        1973      Master       Together     NaN        0         0   
128    8268        1961         PhD        Married     NaN        0         1   
133    1295        1963  Graduation        Married     NaN        0         1   
312    2437        1989  Graduation        Married     NaN        0         0   
319    2863        1970  Graduation         Single     NaN        1         2   
1379  10475        1970      Master       Together     NaN        0         1   
1382   2902        1958  Graduation       Together     NaN        1         1   
1383   4345        1964    2n Cycle         Single     NaN        1         1   
1386   3769        1972         PhD       Together     NaN        1         0   
2059   7187        1969      Master       Together     NaN        1         1   
2061   1612        1981         PhD         Single     NaN        1         0   
2078   5079        1971  Graduation        Married     NaN        1         1   
2079  10339        1954      Master       Together     NaN        0         1   
2081   3117        1955  Graduation         Single     NaN        0         1   
2084   5250        1943      Master          Widow     NaN        0         0   
2228   8720        1978    2n Cycle       Together     NaN        0         0   

     Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
10    2013-11-15       11         5  ...                  7             0   
27    2013-02-20       19         5  ...                  1             0   
43    2013-11-05       80        81  ...                  2             0   
48    2014-01-01       96        48  ...                  6             0   
58    2013-06-17       57        11  ...                  6             0   
71    2012-09-14       25        25  ...                  8             0   
90    2012-11-19        4       230  ...                  9             0   
91    2014-05-27       45         7  ...                  7             0   
92    2013-11-23       87       445  ...                  1             0   
128   2013-07-11       23       352  ...                  6             0   
133   2013-08-11       96       231  ...                  4             0   
312   2013-06-03       69       861  ...                  3             0   
319   2013-08-23       67       738  ...                  7             0   
1379  2013-04-01       39       187  ...                  5             0   
1382  2012-09-03       87        19  ...                  5             0   
1383  2014-01-12       49         5  ...                  7             0   
1386  2014-03-02       17        25  ...                  7             0   
2059  2013-05-18       52       375  ...                  3             0   
2061  2013-05-31       82        23  ...                  6             0   
2078  2013-03-03       82        71  ...                  8             0   
2079  2013-06-23       83       161  ...                  6             0   
2081  2013-10-18       95       264  ...                  7             0   
2084  2013-10-30       75       532  ...                  1             0   
2228  2012-08-12       53        32  ...                  0             0   

      AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedC

Correlación lineal de esta variable con otras del dataset para ver si se puede ser más preciso a la hora de reemplazar el valor de los nulos por algo más elaborado que no sea la media o mediana de esa variable ('Income') y que aporte más información al estudio.

In [331]:
corr_matrix = df.corr(numeric_only=True)
abs(corr_matrix['Income']).sort_values(ascending=False)

Income                 1.000000
NumCatalogPurchases    0.589162
MntMeatProducts        0.584633
MntWines               0.578650
NumWebVisitsMonth      0.553088
NumStorePurchases      0.529362
MntSweetProducts       0.440744
MntFishProducts        0.438871
MntFruits              0.430842
Kidhome                0.428669
NumWebPurchases        0.387878
AcceptedCmp5           0.335943
MntGoldProds           0.325916
AcceptedCmp1           0.276820
AcceptedCmp4           0.184400
Year_Birth             0.161791
Response               0.133047
AcceptedCmp2           0.087545
NumDealsPurchases      0.083101
Complain               0.027225
Teenhome               0.019133
AcceptedCmp3           0.016174
ID                     0.013095
Recency                0.003970
Z_CostContact               NaN
Z_Revenue                   NaN
Name: Income, dtype: float64

Parece que correlaciona alto con varias variables:
- NumCatalogPurchases
- MntMeatProducts
- MntWines

In [332]:
df.loc[df['Income'].isna(),['NumCatalogPurchases','MntMeatProducts','MntWines']]

NumCatalogPurchases  MntMeatProducts  MntWines
10                      0                6         5
27                      0                3         5
43                      3               50        81
48                      1               48        48
58                      0               22        11
71                      0               43        25
90                      2              192       230
91                      0                8         7
92                      4              359       445
128                     1               27       352
133                     5              196       231
312                     5              461       861
319                     3              172       738
1379                    2               65       187
1382                    0               12        19
1383                    0                9         5
1386                    0               13        25
2059                   10               48       375
2061                    0               15        23
2078                    1               16        71
2079                    1               22       161
2081                    1               21       264
2084                    5              490       532
2228                    0             1607        32

In [333]:
df.groupby(['NumCatalogPurchases'])['Income'].describe().round(2)

count       mean       std       min        25%  \
NumCatalogPurchases                                                    
0                    576.0   32898.57  15476.72    1730.0   24451.75   
1                    492.0   41412.80  31350.10    7144.0   31353.00   
2                    274.0   54216.88  14822.29    7500.0   46098.00   
3                    182.0   64390.93  11398.39   32727.0   57203.25   
4                    181.0   67617.66  11656.03   32632.0   60689.00   
5                    137.0   68652.28  11797.09   36262.0   62204.00   
6                    128.0   74335.35  11343.57   42835.0   67445.00   
7                     79.0   76627.33  10105.03   49118.0   70144.00   
8                     55.0   71385.73  13136.96   36130.0   65542.50   
9                     42.0   77524.40  10446.52   56796.0   68682.00   
10                    47.0   75040.53  11129.49   48948.0   68795.00   
11                    19.0   77959.53  10680.07   49476.0   74071.00   
22                     1.0  157243.00       NaN  157243.0  157243.00   
28                     3.0  106798.67  90389.69    2447.0   79796.50   

                          50%        75%       max  
NumCatalogPurchases                                 
0                     31904.0   38875.75  162397.0  
1                     41197.5   49559.25  666666.0  
2                     55185.0   63759.00   90687.0  
3                     64258.0   71764.50   95529.0  
4                     67369.0   75437.00   93404.0  
5                     69930.0   76081.00   94871.0  
6                     75739.5   82253.50   98777.0  
7                     78093.0   83331.50   96876.0  
8                     71853.0   78078.50  105471.0  
9                     79416.0   83149.50  102692.0  
10                    77766.0   83350.00   93027.0  
11                    79800.0   83184.50   96843.0  
22                   157243.0  157243.00  157243.0  
28                   157146.0  158974.50  160803.0

In [334]:
df.groupby(['MntMeatProducts'])['Income'].describe().round(2)

count       mean        std       min        25%       50%  \
MntMeatProducts                                                               
0                  1.0   11448.00        NaN   11448.0   11448.00   11448.0   
1                 14.0   31343.79   37491.26    3502.0   17537.25   21136.5   
2                 30.0   34283.67   24776.43   14421.0   24343.75   31508.0   
3                 39.0   32221.92   18977.96    1730.0   20130.00   31859.0   
4                 30.0   30344.23   12305.13    7500.0   22307.50   26557.5   
...                ...        ...        ...       ...        ...       ...   
984                1.0   80695.00        NaN   80695.0   80695.00   80695.0   
1582               1.0  157243.00        NaN  157243.0  157243.00  157243.0   
1607               0.0        NaN        NaN       NaN        NaN       NaN   
1622               1.0  160803.00        NaN  160803.0  160803.00  160803.0   
1725               2.0   79796.50  109388.71    2447.0   41121.75   79796.5   

                       75%       max  
MntMeatProducts                       
0                 11448.00   11448.0  
1                 34052.00  153924.0  
2                 37163.50  156924.0  
3                 41298.50  113734.0  
4                 39888.25   54466.0  
...                    ...       ...  
984               80695.00   80695.0  
1582             157243.00  157243.0  
1607                   NaN       NaN  
1622             160803.00  160803.0  
1725             118471.25  157146.0  

[558 rows x 8 columns]

Se podría utlizar 'NumCatalogPurchases' y rellenar los valores nulos por la mediana.

## Valores duplicados

In [335]:
df.duplicated().sum()

0

## Primer vistazo variables numéricas

In [336]:
df.describe().T.round(2)

count      mean       std     min       25%      50%  \
ID                   2240.0   5592.16   3246.66     0.0   2828.25   5458.5   
Year_Birth           2240.0   1968.81     11.98  1893.0   1959.00   1970.0   
Income               2216.0  52247.25  25173.08  1730.0  35303.00  51381.5   
Kidhome              2240.0      0.44      0.54     0.0      0.00      0.0   
Teenhome             2240.0      0.51      0.54     0.0      0.00      0.0   
Recency              2240.0     49.11     28.96     0.0     24.00     49.0   
MntWines             2240.0    303.94    336.60     0.0     23.75    173.5   
MntFruits            2240.0     26.30     39.77     0.0      1.00      8.0   
MntMeatProducts      2240.0    166.95    225.72     0.0     16.00     67.0   
MntFishProducts      2240.0     37.53     54.63     0.0      3.00     12.0   
MntSweetProducts     2240.0     27.06     41.28     0.0      1.00      8.0   
MntGoldProds         2240.0     44.02     52.17     0.0      9.00     24.0   
NumDealsPurchases    2240.0      2.33      1.93     0.0      1.00      2.0   
NumWebPurchases      2240.0      4.08      2.78     0.0      2.00      4.0   
NumCatalogPurchases  2240.0      2.66      2.92     0.0      0.00      2.0   
NumStorePurchases    2240.0      5.79      3.25     0.0      3.00      5.0   
NumWebVisitsMonth    2240.0      5.32      2.43     0.0      3.00      6.0   
AcceptedCmp3         2240.0      0.07      0.26     0.0      0.00      0.0   
AcceptedCmp4         2240.0      0.07      0.26     0.0      0.00      0.0   
AcceptedCmp5         2240.0      0.07      0.26     0.0      0.00      0.0   
AcceptedCmp1         2240.0      0.06      0.25     0.0      0.00      0.0   
AcceptedCmp2         2240.0      0.01      0.11     0.0      0.00      0.0   
Complain             2240.0      0.01      0.10     0.0      0.00      0.0   
Z_CostContact        2240.0      3.00      0.00     3.0      3.00      3.0   
Z_Revenue            2240.0     11.00      0.00    11.0     11.00     11.0   
Response             2240.0      0.15      0.36     0.0      0.00      0.0   

                          75%       max  
ID                    8427.75   11191.0  
Year_Birth            1977.00    1996.0  
Income               68522.00  666666.0  
Kidhome                  1.00       2.0  
Teenhome                 1.00       2.0  
Recency                 74.00      99.0  
MntWines               504.25    1493.0  
MntFruits               33.00     199.0  
MntMeatProducts        232.00    1725.0  
MntFishProducts         50.00     259.0  
MntSweetProducts        33.00     263.0  
MntGoldProds            56.00     362.0  
NumDealsPurchases        3.00      15.0  
NumWebPurchases          6.00      27.0  
NumCatalogPurchases      4.00      28.0  
NumStorePurchases        8.00      13.0  
NumWebVisitsMonth        7.00      20.0  
AcceptedCmp3             0.00       1.0  
AcceptedCmp4             0.00       1.0  
AcceptedCmp5             0.00       1.0  
AcceptedCmp1             0.00       1.0  
AcceptedCmp2             0.00       1.0  
Complain                 0.00       1.0  
Z_CostContact            3.00       3.0  
Z_Revenue               11.00      11.0  
Response                 0.00       1.0

Conclusión. Se observan valores inusuales en:
- 'Year_Birth', con valor mínimo de 1893 siendo el máximo 1996 (personas con 103 años)
- 'Income', con un valor máximo de 666.666$ siendo el 75% de 68.522$
- 'MntMeatProducts', con gastos máximos de 1.725$ cuando el 75% está en 232$.

Revisión variable 'Year_Birth'

In [337]:
df.loc[df['Year_Birth'] < 1941]

ID  Year_Birth Education Marital_Status   Income  Kidhome  Teenhome  \
192    7829        1900  2n Cycle       Divorced  36640.0        1         0   
239   11004        1893  2n Cycle         Single  60182.0        0         1   
339    1150        1899       PhD       Together  83532.0        0         0   
1950   6663        1940       PhD         Single  51141.0        0         0   

     Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
192   2013-09-26       99        15  ...                  5             0   
239   2014-05-17       23         8  ...                  4             0   
339   2013-09-26       36       755  ...                  1             0   
1950  2013-07-08       96       144  ...                  5             0   

      AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
192              0             0             0             0         1   
239              0             0             0             0         0   
339              0             1             0             0         0   
1950             0             0             0             0         0   

      Z_CostContact  Z_Revenue  Response  
192               3         11         0  
239               3         11         0  
339               3         11         0  
1950              3         11         0  

[4 rows x 29 columns]

Conclusión: Hay tres personas únicamente con edades cercanas a los 100 años, siendo la anterior de 56 años. ¿?

Revisión variable 'Income'

In [338]:
df.loc[df['Income'] > 100000]

ID  Year_Birth   Education Marital_Status    Income  Kidhome  \
124    7215        1983  Graduation         Single  101970.0        0   
164    8475        1973         PhD        Married  157243.0        0   
203    2798        1977         PhD       Together  102160.0        0   
252   10089        1974  Graduation       Divorced  102692.0        0   
617    1503        1976         PhD       Together  162397.0        1   
646    4611        1970  Graduation       Together  105471.0        0   
655    5555        1975  Graduation       Divorced  153924.0        0   
687    1501        1982         PhD        Married  160803.0        0   
1300   5336        1971      Master       Together  157733.0        1   
1653   4931        1977  Graduation       Together  157146.0        0   
1898   4619        1945         PhD         Single  113734.0        0   
2132  11181        1949         PhD        Married  156924.0        0   
2233   9432        1977  Graduation       Together  666666.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
124          0  2013-03-12       69       722  ...                  2   
164          1  2014-03-01       98        20  ...                  0   
203          0  2012-11-02       54       763  ...                  4   
252          0  2013-04-05        5       168  ...                  2   
617          1  2013-06-03       31        85  ...                  1   
646          0  2013-01-21       36      1009  ...                  3   
655          0  2014-02-07       81         1  ...                  0   
687          0  2012-08-04       21        55  ...                  0   
1300         0  2013-06-04       37        39  ...                  1   
1653         0  2013-04-29       13         1  ...                  1   
1898         0  2014-05-28        9         6  ...                  1   
2132         0  2013-08-29       85         2  ...                  0   
2233         0  2013-06-02       23         9  ...                  6   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
124              0             1             1             1             0   
164              0             0             0             0             0   
203              0             1             1             1             0   
252              0             1             1             1             1   
617              0             0             0             0             0   
646              0             0             1             1             0   
655              0             0             0             0             0   
687              0             0             0             0             0   
1300             0             0             0             0             0   
1653             0             0             0             0             0   
1898             0             0             0             0             0   
2132             0             0             0             0             0   
2233             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
124          0              3         11         1  
164          0              3         11         0  
203          0              3         11         1  
252          0              3         11         1  
617          0              3         11         0  
646          0              3         11         1  
655          0              3         11         0  
687          0              3         11         0  
1300         0              3         11         0  
1653         0              3         11         0  
1898         0              3         11         0  
2132         0              3         11         0  
2233         0              3         11         0  

[13 rows x 29 columns]

In [339]:
df.loc[df['Income'] < 12000]

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
9      5899        1950         PhD       Together   5648.0        1   
11      387        1976       Basic        Married   7500.0        0   
21     5376        1979  Graduation        Married   2447.0        1   
30     6864        1989      Master       Divorced  10979.0        0   
44     2139        1975      Master        Married   7500.0        1   
46     9909        1996    2n Cycle        Married   7500.0        0   
136    5342        1976       Basic       Divorced   9548.0        1   
238    7297        1973  Graduation         Single   7500.0        1   
385     948        1971  Graduation         Single  10245.0        1   
428    5987        1989    2n Cycle       Together  10404.0        1   
439     456        1986    2n Cycle        Married   7500.0        1   
724    4692        1976  Graduation        Married   7500.0        1   
774    4303        1957         PhD       Together   6835.0        0   
862    9553        1987  Graduation        Married   7500.0        0   
981    3955        1965  Graduation       Divorced   4861.0        0   
1042  10749        1991  Graduation         Single   8028.0        0   
1138  10617        1989      Master       Divorced  10979.0        0   
1152  10710        1979  Graduation        Married   7500.0        0   
1220  10395        1986       Basic         Single   8940.0        1   
1245   6862        1971  Graduation       Divorced   1730.0        0   
1246   4136        1992       Basic         Single   7500.0        1   
1299  10641        1978  Graduation         Single   7500.0        1   
1328   4246        1982      Master         Single   6560.0        0   
1524  11110        1973  Graduation         Single   3502.0        1   
1536    451        1989  Graduation         Single   9255.0        1   
1704   7349        1973       Basic       Together   9722.0        1   
1806   6237        1966         PhD         Single   7144.0        0   
1846   9931        1963         PhD        Married   4023.0        1   
1861  10241        1975    2n Cycle       Divorced  11448.0        0   
1975  10311        1969  Graduation        Married   4428.0        0   
2029  10001        1985    2n Cycle       Together   7500.0        1   
2195   2980        1952      Master         Single   8820.0        1   
2214   9303        1976  Graduation        Married   5305.0        0   
2222  10659        1979    2n Cycle       Together   7500.0        1   
2230   7004        1984  Graduation         Single  11012.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
9            1  2014-03-13       68        28  ...                 20   
11           0  2012-11-13       59         6  ...                  8   
21           0  2013-01-06       42         1  ...                  1   
30           0  2014-05-22       34         8  ...                  5   
44           0  2013-10-02       19         3  ...                  5   
46           0  2012-11-09       24         3  ...                  9   
136          0  2012-08-08       31         0  ...                  8   
238          0  2013-12-26       54         5  ...                  7   
385          0  2013-05-15       32         4  ...                  5   
428          0  2014-05-04       52         2  ...                  5   
439          0  2013-02-07       96         1  ...                  8   
724          0  2012-08-01       19         7  ...                  9   
774          1  2012-12-08       76       107  ...                 20   
862          0  2013-01-09       94         0  ...                  6   
981          0  2014-06-22       20         2  ...                 14   
1042         0  2012-09-18       62        73  ...                 19   
1138         0  2014-05-22       34         8  ...                  5   
1152         1  2012-08-29       61         5  ...                  8   
1220         0  2012-08-22   

Conclusión:  
+ Hay una sola persona que tiene unos ingresos de 666.666$ lo que considero un outlier que eliminaré.   
+ Los valores de ingresos más bajos parece que hay más, si bien son muchos los que están por debajo de 12000$. Los dejaré.

Revisión variable 'MntMeatProducts'

In [340]:
# Unusual values in 'MntMeatProducts' variable
df.loc[df['MntMeatProducts'] > 900]

ID  Year_Birth   Education Marital_Status    Income  Kidhome  \
21     5376        1979  Graduation        Married    2447.0        1   
76     1859        1972      Master        Married   75251.0        0   
126   11071        1984  Graduation         Single   79607.0        0   
140     821        1992      Master         Single   92859.0        0   
164    8475        1973         PhD        Married  157243.0        0   
456    4947        1966    2n Cycle         Single   89572.0        0   
546   10955        1964  Graduation          Widow   85620.0        0   
561    3179        1980  Graduation         Single   81741.0        0   
687    1501        1982         PhD        Married  160803.0        0   
845    5848        1970  Graduation       Together   81205.0        0   
966   10489        1973  Graduation        Married   92955.0        0   
1031   9220        1971  Graduation         Single   91700.0        0   
1179   5735        1991      Master         Single   90638.0        0   
1213    697        1969  Graduation       Together   74918.0        0   
1259   2147        1969  Graduation       Together   76653.0        0   
1358   1065        1963  Graduation         Single   80695.0        0   
1572   5350        1991      Master         Single   90638.0        0   
1653   4931        1977  Graduation       Together  157146.0        0   
1713   7396        1966  Graduation        Married   80398.0        0   
1721   1172        1983         PhD        Married   92491.0        0   
1808   1619        1956  Graduation        Married   90369.0        0   
1958   2109        1990  Graduation         Single   96843.0        0   
2010   2002        1958  Graduation       Divorced   71964.0        0   
2109   3104        1961  Graduation         Single   82332.0        0   
2228   8720        1978    2n Cycle       Together       NaN        0   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
21           0  2013-01-06       42         1  ...                  1   
76           0  2012-08-27       34       721  ...                  5   
126          0  2013-04-21       37       450  ...                  1   
140          0  2012-10-19       46       962  ...                  2   
164          1  2014-03-01       98        20  ...                  0   
456          0  2012-09-15       44       606  ...                  4   
546          0  2014-01-07       68       416  ...                  1   
561          0  2013-06-16       75       249  ...                  2   
687          0  2012-08-04       21        55  ...                  0   
845          0  2014-01-13       43       724  ...                  2   
966          0  2013-08-19       35       693  ...                  2   
1031         0  2013-01-17       58       576  ...                  3   
1179         0  2014-02-13       29      1156  ...                  1   
1213         0  2013-03-12       78       972  ...                  3   
1259         0  2013-08-16       91       736  ...                  2   
1358         0  2013-12-29       85       557  ...                  2   
1572         0  2014-02-13       29      1156  ...                  1   
1653         0  2013-04-29       13         1  ...                  1   
1713         0  2012-11-10       92       342  ...                  3   
1721         0  2013-05-01       71       979  ...                  3   
1808         0  2014-04-28        2       292  ...                  1   
1958         0  2013-04-23       60       448  ...                  2   
2010         0  2013-05-20       81       587  ...                  3   
2109         0  2012-09-17       89       830  ...                  2   
2228         0  2012-08-12       53        32  ...                  0   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
21               0             0             0             0             0   
76               0             0             0         

Conclusión: Parece que hay varias personas que están por encima de 900$. 

___

## Resumen de conclusiones iniciales.

**Tamaño**:
- 2240 entradas
- 29 columnas

**Variables**:
- **ID**: 100% cardinalidad, podría ser el Índice. -> **ACCIÓN: Cambiar a índice**
- **Dt_Customer**: tiene tipo 'objeto' y podría ser 'datetime'. -> **ACCIÓN: cambiar tipo**.
- **Z_CostContact** y **Z_Revenue**: tienen solo un valor cada uno. -> **ACCIÓN: eliminar columnas**
- **'Education', 'Marital_Status','AcceptedCmp3',
    'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2',
    'Complain'**: tienen tipo 'int' y podrían ser 'categoría'. -> **ACCIÓN: cambiar tipo**
- **Marital_Status**: tiene dos valores llamados ‘Absurd’ y ‘YOLO’ que no tienen sentido.
    - YOLO: será eliminado -> **ACCIÓN: eliminar**
    - Absurd: será incluido en una variable única llamada ‘otros’. -> **ACCIÓN: cambiar a 'otros'**
- **Response**: podría ser la variable objetivo.

**Valores nulos**:
- **Income**: tiene 24 valores nulos. -> **ACCIÓN: Rellenar valores nulos** con la mediana de 'NumCatalogPurchases'

**Valores atípicos**:
- Year_Birth (Personas que tienen más de 90 años; anteriormente, los datos indicaban una edad promedio de 56 años):
    - Número de ID: 
        - 7829
        - 11004
        - 1150
- Income (666.666$):
    - Número de ID: 9432


___

### Posibles nuevas columnas:
- Edad
- Antigüedad del cliente
- Miembros del hogar
- Gasto total
- Número de ofertas aceptadas
- Número total de compras
- Gasto promedio (Gasto total/Número total de compras)

### Posible discretizacion de variables
- Income
- Edad

In [341]:
df.to_csv('./data/data_preprocessing.csv')